src/core_plugins/input_control_vis/public/components/editor/control_editor.js

```diff
@@ -0,0 +1,111 @@
+import classNames from 'classnames';
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+
+export class ControlEditor extends Component {
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      isEditorVisible: true
+    };
+  }
+
+  handleToggleControlVisibility() {
+    this.setState({
+      isEditorVisible: !this.state.isEditorVisible
+    });
+  }
+
+  render() {
+    let editor;
+    if (this.state.isEditorVisible) {
+      editor = (
+        <div>
+          <div className="kuiSideBarFormRow">
+            <label className="kuiSideBarFormRow__label">
+              Label
+            </label>
+            <div className="kuiSideBarFormRow__control kuiFieldGroupSection--wide">
+              <input
+                className="kuiTextInput"
+                type="text"
+                value={this.props.controlParams.label}
+                onChange={this.props.handleLabelChange}
+              />
+            </div>
+          </div>
+
+          {this.props.children}
+        </div>
+      );
+    }
+
+    const visibilityToggleClasses = classNames('fa', {
+      'fa-caret-right': !this.state.isEditorVisible,
+      'fa-caret-down': this.state.isEditorVisible
+    });
+
+    let controlTitle = `${this.props.controlParams.type}: ${this.props.controlIndex}`;
+    if (this.props.controlParams.label) {
+      controlTitle = `${this.props.controlParams.type}: ${this.props.controlParams.label}`;
+    } else if (this.props.controlParams.fieldName) {
+      controlTitle = `${this.props.controlParams.type}: ${this.props.controlParams.fieldName}`;
```

Comment:
Yes, I would apply that to props as well. In addition to the argument of primitives being easier to compare for `shouldComponentUpdate` checks, passing separate props also documents the external interface of the component better an an opaque "params" object. In general I found that many clean-code principles for function arguments are also applicable to component props.

Reply:
I don't think props needs to be as strict as state. I can think of several use-cases for objects and arrays in `props`, e.g. let's say you have a `Clusters` component that renders clusters: It should be able to receive an array of cluster objects, and maybe even do something like:

```js
return <ul>
  {
    clusters.map(cluster => <li><Cluster cluster={ cluster } /></li>)
  }
</ul>
```

Also, for large objects having to list everything in the component props would be a little heavy, I think.

When using objects or arrays, I think we should prefer to be more specific in the proptypes though. E.g. for an object we should often use `PropTypes.shape` (unless purely passing through to another component, where `PropTypes.object` might be okey as the underlying component would do a `PropTypes.shape`).

In the end it feels to me like a "it feels like we should use an object here" vs "it feels like we should use primitives here", which would be more case-by-case rather than strictly defined. (Might be difficult to be super-consistent with that, though)

Does that make sense?